File: testerum-backend/testerum-runner/testerum-runner-cmdline/src/main/assembly/report_templates/common/util/RunnerPropertiesParser.ts
export class RunnerPropertiesParser {

    private static readonly PROPERTIES_SPLITTER    = /(?<!\\),/g;
    private static readonly PROPERTY_PAIR_SPLITTER = /(?<!\\)=/g;

    public static parse(propertiesString: string): {[key: string]: string} {
        if (!propertiesString || propertiesString === "") {
            return {};
        }

        const result = {};

        const propertyPairs: Array<string> = propertiesString.split(this.PROPERTIES_SPLITTER);
        for (const propertyPair of propertyPairs) {
            const propertyPairParts: Array<string> = propertyPair.split(this.PROPERTY_PAIR_SPLITTER);

            if (propertyPairParts.length == 1) {
                const key = this.unescape(propertyPairParts[0]);

                result[key] = ""
            } else if (propertyPairParts.length == 2) {
                const key = this.unescape(propertyPairParts[0]);
                const value = this.unescape(propertyPairParts[1]);

                result[key] = value
            } else {
                throw Error(
                    `[${propertiesString}] has invalid format" +
                    ": the pair [${propertyPair}] should contain an '=' character that separates the key and the value`
                )
            }
        }

        return result;
    }

    private static unescape(text: string): string {
        return text.replace("\\=", "=")
            .replace("\\,", ",")
    }

}

File: testerum-backend/testerum-runner/testerum-runner-api/src/main/kotlin/com/testerum/runner/cmdline/report_type/marshaller/properties/RunnerPropertiesParser.kt
package com.testerum.runner.cmdline.report_type.marshaller.properties

object RunnerPropertiesParser {

    private val PROPERTIES_SPLITTER = Regex("""(?<!\\),""")
    private val PROPERTY_PAIR_SPLITTER = Regex("""(?<!\\)=""")

    fun parse(propertiesString: String): Map<String, String> {
        if (propertiesString.isEmpty()) {
            return emptyMap()
        }

        val result = LinkedHashMap<String, String>()

        val propertyPairs = PROPERTIES_SPLITTER.split(propertiesString)
        for (propertyPair in propertyPairs) {
            val propertyPairParts = PROPERTY_PAIR_SPLITTER.split(propertyPair)

            if (propertyPairParts.size == 1) {
                val key = propertyPairParts[0].unescape()

                result[key] = ""
            } else if (propertyPairParts.size == 2) {
                val key = propertyPairParts[0].unescape()
                val value = propertyPairParts[1].unescape()

                result[key] = value
            } else {
                throw IllegalArgumentException(
                        "[$propertiesString] has invalid format" +
                                ": the pair [$propertyPair] should contain an '=' character that separates the key and the value"
                )
            }
        }

        return result
    }

    private fun String.unescape(): String {
        return this.replace("\\=", "=")
                .replace("\\,", ",")
    }

}
